diff --git a/llvm/lib/Support/Jobserver.cpp b/llvm/lib/Support/Jobserver.cpp
index c75e583301d0..91918786033a 100644
--- a/llvm/lib/Support/Jobserver.cpp
+++ b/llvm/lib/Support/Jobserver.cpp
@@ -158,7 +158,7 @@ public:
   bool isValid() const { return IsInitialized; }
 
 private:
-#if defined(LLVM_ON_UNIX)
+#if defined(LLVM_ON_UNIX) && !defined(__wasi__)
   int ReadFD = -1;
   int WriteFD = -1;
   std::string FifoPath;
@@ -172,8 +172,8 @@ private:
 };
 } // namespace llvm
 
 // Include the platform-specific parts of the class.
-#if defined(LLVM_ON_UNIX)
+#if defined(LLVM_ON_UNIX) && !defined(__wasi__)
 #include "Unix/Jobserver.inc"
 #elif defined(_WIN32)
 #include "Windows/Jobserver.inc"
@@ -227,7 +227,7 @@ JobserverClient *JobserverClient::getInstance() {
     }
 
     if (Config.TheMode == JobserverConfig::PosixPipe) {
-#if defined(LLVM_ON_UNIX)
+#if defined(LLVM_ON_UNIX) && !defined(__wasi__)
       if (!areFdsValid(Config.PipeFDs.Read, Config.PipeFDs.Write)) {
         errs() << "Warning: failed to create jobserver client due to invalid "
                   "Pipe FDs in MAKEFLAGS environment variable\n";
